### ISSUE TITLE
Moving the Gallery app to the Appframework

### ISSUE THREAD

**Issue by oparoz** [owncloud/gallery#151]
### TL;DR
The Gallery app has been rewritten as an Appframework app and I kept the same UI and logic.
Now would be a good time to switch over to that version so that we can move forward with development.
### Introduction
Back in December I created a PR to add support for all the media types which ownCloud supports via its config file: https://github.com/oparoz/gallery/tree/support-more-media-types
It was suggested to me that using the AppFramework to implement such changes would be a better idea as it's the official way to write OC apps and because it facilitates testing.

### Requirements
* Get rid of obsolete code
* Try to rely exclusively on public APIs and the Appframework
* Keep the UI
* Detect and support all media types configured in OC via its config file
* Have at least one release compatible with OC 7

### Current status
- [x] Write the app
- [x] Add all current PRs which were validated but not merged
- [x] Include fixes for the Preview Manager which are not yet in OC master
- [ ] Review the PHP. https://github.com/interfasys/galleryplus/issues/11
- [ ] Review the JS https://github.com/interfasys/galleryplus/issues/12
- [x] Migrate the app to the owncloud domain
- [ ] Write unit tests
- [ ] Write integration tests

### Next steps
In order
- [x] App moves to the owncloud domain as galleryplus since the Pictures app is already taking the gallery space (unless you want to move that app to pictures). @karlitschek @DeepDiver1975 
- [x] App name in oC should be Gallery and not Gallery+. @oparoz
- [ ] Translations should be setup using the 'galleryplus' id since the Pictures app is already using 'gallery'. @oparoz, with the help of @DeepDiver1975 to set up Transiflex
- [ ] Give the app (in the app store) the approved status, so that it shows up in the app manager on oC 8.0 @DeepDiver1975 ?

cc @icewind1991 @georgehrke @libasys @jancborchardt

**Comment by jospoortvliet** [comment]
@karlitschek @DeepDiver1975 something awesome to review for as soon as oC8 is out ;-)

**Comment by karlitschek** [comment]
nice.
We should move this over in the owncloud organisation if @oparoz agrees. I´m not sure if this should replace the current gallery repository. Maybe do a parallel repo for now. For example "galleryplus".
Then we can see if we switch over to this new app in the next release. The requirement for that is that we don´t have any feature or bug regressions.

**Comment by oparoz** [comment]
@karlitschek - Yes, I agree to move this over to owncloud. The only problem I see with having 2 parallel versions is regarding translations. I've kept the same ID so that people didn't need to add the same translations twice, but I've added some new strings which are not yet in the original repository's l10n folder. I don't know if someone else has been in the same situation before. Should we add the strings to the original project even though they're not being used yet and get Transifex to push changes to both repositories? Or can we symlink folders?

Also, I can't move the repository myself it seems as only an owncloud admin can make that change. Maybe this can be initiated from owncloud? If not, then I'll need to figure out how to push the local repository to a fresh one without losing any commits.

**Comment by jospoortvliet** [comment]
@karlitschek can make the move. not sure about translations...

**Comment by karlitschek** [comment]
@DeepDiver1975 Can you help with moving the repository.
@oparoz Let´s keep in separate repositories for now with separate issues and separate translations. 
Than we can offer two galleries for some time and in a second step decide which one will be the default.
I hope this works for everyone.

**Comment by oparoz** [comment]
Works for me.
One more thing to bear in mind. This project only has code commits starting from the 5th of January, as a new project. Not sure if it's a problem or not. The original project can be used as a reference, but you may want full commit history in this one, even if there is a big change in the middle. Up to you @karlitschek and @icewind1991 .

**Comment by jancborchardt** [comment]
Then it might be best to directly call it »Media«. As in the future videos should also be directly in there. Calling it »Gallery Plus« just seems strange and it will be difficult to transition to a proper name if we decide to keep that one.

**Comment by jancborchardt** [comment]
@icewind1991 it would be cool to get your input and verdict on this since you are basically the main Gallery developer. Are you cool with the change?

**Comment by oparoz** [comment]
@jancborchardt I'm fine with calling it "Media" and I'm looking forward to the new app icon you're going to design ;)

**Comment by butonic** [comment]
Users might expect to find music in an app called  "Media" ... not sure about the name ... I'd make "Pictures" and "Videos" separate apps ... my 2 cents

**Comment by DeepDiver1975** [comment]
media is kind of burned - the old music player was called media/media player

**Comment by jancborchardt** [comment]
Then we should call it »Gallery« and actually make a conscious decision to replace it. We’re not making separate apps for pictures and videos simply because we messed up names cause of legacy reasons.

**Comment by jospoortvliet** [comment]
@jancborchardt fully agreed. Let's just name it Gallery, that's a great name. That this is a much improved version is only good, right?

**Comment by oparoz** [comment]
"Gallery" actually works for visual media. People can still install a separate video player if they want.

**Comment by karlitschek** [comment]
I agree. Good idea to not call it Pictures so that users who have both apps installed can see the difference.

**Comment by jospoortvliet** [comment]
I'm guessing this won't be for 8.1, or will it?

**Comment by oparoz** [comment]
### Next steps
In order
- [ ] App moves to the owncloud domain as galleryplus since the Pictures app is already taking the gallery space (unless you want to move that app to pictures). @karlitschek @DeepDiver1975 
- [ ] App name in oC should be Gallery and not Gallery+. @oparoz
- [ ] Translations should be setup using the 'galleryplus' id since the Pictures app is using 'gallery' already. @oparoz, with the help of @DeepDiver1975 to set up Transiflex
- [ ] Give the app (in the app store) the approved status, so that it shows up in the app manager on oC 8.0 @DeepDiver1975 ?

Some buttons have been added to the UI, but I don't think it's a concern right now unless you want to ship the app as part of the oC distribution, like you do for the **currently broken Pictures app**.
[The demo](https://oc8demo.interfacloud.com/index.php/apps/galleryplus/s/4jjwPv5E21l8Lnk) contains the latest UI.

Most of the work needs to be done by me, so I think it's feasible to get it done on time for the launch of 8.1. That would provide users with a working alternative Pictures app and not leave them stranded like at the launch of 8.0.

**Comment by jancborchardt** [comment]
@oparoz that sounds good.

What do you think about having a call or something about the UI changes, like the added buttons? I want to improve the look for some of them to make them look more like ownCloud, but we also need to generally discuss if all of them are needed (like the background switcher and the info button).

**Comment by oparoz** [comment]
@jancborchardt I'm up for anything which get things moving forward :).
You can submit PRs to modify buttons since it's not like I was trying to introduce a new design :D. I'm not particularly attached to them.

And I'm happy to discuss about the rest. The app can now enable features on-demand, so it's less of a problem to remove things which the team feels are unnecessary, but which I need. The important thing is that it meets oC's requirements.

**Comment by jancborchardt** [comment]
@oparoz great stuff! :) I will see which pull requests I can come up with then.

**Comment by jeffhoffman13** [comment]
This looks good but missing the slide show portion..

**Comment by oparoz** [comment]
@jeffhoffman13 - The slideshow is there. If it's not showing up for you, please [follow these instructions](https://github.com/interfasys/galleryplus/blob/master/README.md#bugs) and open a ticket

**Comment by jeffhoffman13** [comment]
After generating all the images for a given folder the slide show functions as expected.  ;)

**Comment by DeepDiver1975** [comment]
> App moves to the owncloud domain as galleryplus since the Pictures app is already taking the gallery space (unless you want to move that app to pictures). @karlitschek @DeepDiver1975

@oparoz please transfer the repo to my github id - I can then move it to the owncloud org. Once in the org repo we can rename it as we like.

**Comment by DeepDiver1975** [comment]
.... sorry for not taking care of this the past weeks - I have been busy ...

**Comment by oparoz** [comment]
@DeepDiver1975 - I have the same problem when trying to push to your domain, so if I make you an admin, I may as well push it directly to the owncloud domain, no?

**Comment by DeepDiver1975** [comment]
Hmm ... we can try that as well .... I honestly don't remember what the right process is.

Let's give it a try - make me an admin.  Thx

**Comment by oparoz** [comment]
Invite sent and there is some doc here:
https://help.github.com/articles/transferring-a-repository/#transferring-between-organizations

**Comment by DeepDiver1975** [comment]
here we go - welcome to the owncloud org !!!
https://github.com/owncloud/galleryplus

**Comment by oparoz** [comment]
Nice! That should help reassure some people who were hesitant to install something from an unknown provider.

Next steps, decide on the name and set up the transiflex.

**Comment by jeffhoffman13** [comment]
This seems to be ok on 8.0.2 ;) and 8.0.3 non-external storage.

Different issue with external storage local attachment not remaining 
after I enter them under admin in 8.0.3.. trying to chase that one down..

Jeff

**Comment by jancborchardt** [comment]
@jeffhoffman13 your reports are appreciated, but please do not use this issue as a catch-all. Instead report to the respective repositories as outlined in the contribution guidelines: https://github.com/owncloud/core/blob/master/CONTRIBUTING.md

**Comment by jancborchardt** [comment]
@oparoz nice! Now can we rename the app to »Gallery«? (That is, the user-facing name.) :)

**Comment by oparoz** [comment]
@jancborchardt - Done and there are now 2 (8.0 & 8.1) version in the app store.

**Comment by jancborchardt** [comment]
@oparoz great stuff! We’re on a good way here. :) Thank you for all the work!

**Comment by DeepDiver1975** [comment]
@jancborchardt I suggest we move on with this tomorrow while we meet for coworking.

**Comment by jancborchardt** [comment]
@DeepDiver1975 yes!

**Comment by jancborchardt** [comment]
@oparoz we will unfortunately need to postpone the whole switch to 8.2 due to several problems:
* renaming the current gallery app / repo will break all release scripts and documentation
* the Gallery+ says it supports IE11, but says nothing about IE8–10. We have to support those though. (Maybe by 8.2 we can drop IE8 support, maybe not)
* there are some design enhancements I need to do – not enough time
* we just don’t have enough time for proper testing so it has a high risk of release-breaking bugs or bad user experience

Hope that’s understandable. For 8.2 then @DeepDiver1975 and I (and hopefully others as well) will work together with you to get Gallery+ the new default ownCloud Gallery app! :)

**Comment by oparoz** [comment]
@jancborchardt - I, personally, don't mind. It's more of a problem for ownCloud users.

* I don't know how much work is involved in updating the release scripts, so I can't comment on that
* Since the interface is similar to Pictures, I don't see updating the doc as being a major task
* The current Pictures app doesn't support any IE version and it's the only app which ships with 8.0, so I don't think that's a valid argument against introducing one which only officially supports IE11, but which might work with older versions
* Design enhancement. If it's a visible feature which needs hiding while waiting for a better design, it shouldn't be a problem to hide it. If there are a lot of things which needs changing, then yes, it would waste too much time trying to hide stuff
* Regarding bugs, it breaks much less often than Pictures which offers a really bad UX (endless spinning wheel when anything wrong happens), but it's true that the introduced advanced features are probably less tested and unit tests are still missing.

What I would suggest is to make it available in the app manager, so that users can quickly switch to it in case Pictures doesn't work for them. Most people will prefer a one click solution vs downloading and manually tracking apps from the app store. I've seen people stay with their broken Pictures simply because they don't trust the app store or don't know how to install apps.

**Comment by jancborchardt** [comment]
@oparoz those points are all true. :) The main thing is simply that we are too far into the release cycle (waayyyy past feature freeze) to now change one of the core apps. Gallery+ is just not tested to a level how Pictures is (by so many people). There might always be bugs.

So yes – Gallery+ will of course already show up in the apps management since you added it to apps.owncloud.com

And everything else we will take care of together after we released 8.1 :)

**Comment by oparoz** [comment]
I agree, it's late in the game and I thought it wouldn't happen before 8.2, so I'm not shocked ;)

Regarding the app store, in 8.1 beta1, only the official apps show app in the app manager. Do you know when that's due to change?

**Comment by jancborchardt** [comment]
@oparoz good question, I also encounter that issue that no apps from the store really show up in my dev installation. ;) It will definitely work when it’s released. cc @LukasReschke

**Comment by deMattin** [comment]
The pictures app is broken and doesn't work with most 8.x installations!
I don't know, which tests have to be done to name an app "tested" - but I wonder, how pictures app (oc8) has done this tests.
Holding back the gallery(+) app for another half of a year isn't a good idea.

And I don't understand it at all.
The apps are separated from core and can be activated and deactivated separately.
It has been the argument to do it like that, that it is more simple to update apps separately.
So why not give the owncloud users the possibility to install this much better an working app in a simple way without using git?
Even if there were some bugs left, that are not seen yet - it's better to give the user a working app which maybe needs some design enhancements than giving the user an app (pictures), that doesn't work at all.
So please rethink the decision.
I'm not able to help with coding but maybe with testing ...

If gallery(+) can be simply activated (without git) in 8.1 and this is only a discussion of fully substituting the pictures app, I precautionary say  sorry, that I don't know or understand, how the new app store or installing of apps in 8.1+ is working ... ;)

**Comment by jospoortvliet** [comment]
@deMattin while I personally would also love to see the galleryPlus app enter sooner rather than later (it can't make the pictures app more broken than it is, to be honest), I'll leave it up to @jancborchardt and @DeepDiver1975 to decide.

I do want to point out that it won't mean a half year delay, we're now on a 3 month release cycle so 8.2 should be out in about 4 months.

**Comment by oparoz** [comment]
Almost there. PR has been created https://github.com/owncloud/gallery/pull/200

**Comment by oparoz** [comment]
Will close this now as the move is happening and users should get the new Gallery for 8.2.
